mlrun/api/db/sqldb/db.py

```diff
@@ -3764,3 +3764,41 @@ def delete_run_notifications(
 
         if commit:
             session.commit()
+
+    def set_run_notifications(
+        self,
+        session: Session,
+        project: str,
+        notifications: typing.List[mlrun.model.Notification],
+        identifier: mlrun.common.schemas.RunIdentifier,
+        **kwargs,
+    ):
+        """
+        Set notifications for a run. This will replace any existing notifications.
+        :param session: SQLAlchemy session
+        :param project: Project name
+        :param notifications: List of notifications to set
+        :param identifier: Run identifier
+        :param kwargs: Ignored additional arguments (for interfacing purposes)
+        """
+        run = self._get_run(session, identifier.uid, project, None)
+        if not run:
+            raise mlrun.errors.MLRunNotFoundError(
+                "Could not find run",
+                project=project,
+                uid=identifier.uid,
+            )
+
+        run.struct.setdefault("spec", {})["notifications"] = [
+            notification.to_dict() for notification in notifications
+        ]
+        self._upsert(session, [run], ignore=True)
+
+        self.delete_run_notifications(session, run_uid=run.uid, project=project)
+        if notifications:
+            self.store_run_notifications(
+                session,
+                notification_objects=notifications,
+                run_uid=run.uid,
+                project=project,
+            )
```

Comment:
can that be a single transaction ?, meaning if one operation failed, the rest rollback?